### PR TITLE
AWS: Fix #2796 - avoid S3 error of "Resetting to invalid mark" by re-creating input stream on retries

### LINKED ISSUE #2796: Sporadic `Resetting to invalid mark` Error
We're trying to use S3FileIO and we're running into the following exception on uploads:

```
java.io.UncheckedIOException: java.io.IOException: Resetting to invalid mark
	at software.amazon.awssdk.utils.FunctionalUtils.asRuntimeException(FunctionalUtils.java:180)
	at software.amazon.awssdk.utils.FunctionalUtils.lambda$safeRunnable$5(FunctionalUtils.java:126)
	at software.amazon.awssdk.utils.FunctionalUtils.invokeSafely(FunctionalUtils.java:140)
	at software.amazon.awssdk.core.sync.RequestBody.lambda$fromInputStream$1(RequestBody.java:122)
	at software.amazon.awssdk.services.s3.internal.handlers.SyncChecksumValidationInterceptor$ChecksumCalculatingStreamProvider.lambda$newStream$0(SyncChecksumValidationInterceptor.java:112)
	at software.amazon.awssdk.utils.FunctionalUtils.lambda$safeSupplier$4(FunctionalUtils.java:108)
	at software.amazon.awssdk.utils.FunctionalUtils.invokeSafely(FunctionalUtils.java:136)
	at software.amazon.awssdk.services.s3.internal.handlers.SyncChecksumValidationInterceptor$ChecksumCalculatingStreamProvider.newStream(SyncChecksumValidationInterceptor.java:112)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage$ClosingStreamProvider.newStream(StreamManagingStage.java:78)
	at java.util.Optional.map(Optional.java:215)
	at software.amazon.awssdk.http.apache.internal.RepeatableInputStreamRequestEntity.getContent(RepeatableInputStreamRequestEntity.java:114)
	at software.amazon.awssdk.http.apache.internal.RepeatableInputStreamRequestEntity.<init>(RepeatableInputStreamRequestEntity.java:89)
	at software.amazon.awssdk.http.apache.internal.impl.ApacheHttpRequestFactory.wrapEntity(ApacheHttpRequestFactory.java:153)
	at software.amazon.awssdk.http.apache.internal.impl.ApacheHttpRequestFactory.createApacheRequest(ApacheHttpRequestFactory.java:133)
	at software.amazon.awssdk.http.apache.internal.impl.ApacheHttpRequestFactory.create(ApacheHttpRequestFactory.java:53)
	at software.amazon.awssdk.http.apache.ApacheHttpClient.toApacheRequest(ApacheHttpClient.java:258)
	at software.amazon.awssdk.http.apache.ApacheHttpClient.prepareRequest(ApacheHttpClient.java:228)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder$NonManagedSdkHttpClient.prepareRequest(SdkDefaultClientBuilder.java:432)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.executeHttpRequest(MakeHttpRequestStage.java:67)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.execute(MakeHttpRequestStage.java:55)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.execute(MakeHttpRequestStage.java:39)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:73)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:77)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:39)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:50)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:64)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:34)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:56)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:80)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:60)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:48)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:31)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:26)
	at software.amazon.awssdk.core.internal.http.AmazonSyncHttpClient$RequestExecutionBuilderImpl.execute(AmazonSyncHttpClient.java:193)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.invoke(BaseSyncClientHandler.java:133)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:159)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$1(BaseSyncClientHandler.java:112)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:167)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:94)
	at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:45)
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:55)
	at software.amazon.awssdk.services.s3.DefaultS3Client.putObject(DefaultS3Client.java:8645)
	at org.apache.iceberg.aws.s3.S3OutputStream.completeUploads(S3OutputStream.java:316)
	at org.apache.iceberg.aws.s3.S3OutputStream.close(S3OutputStream.java:197)
```

This occurs sporadically and can succeed on retries, but it happens consistently.

I was wondering if anyone else has seen this? I was wondering if the RequestBody.fromInputStream could be setting a mark at 128kB, then start reading past it, hit some sort of exception, and try to reset back to the beginning.

https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/RequestBody.java#L131

This would cause the BufferedInputStream to fail created here because it has read past the readlimit of the mark.

https://github.com/apache/iceberg/blob/master/aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java#L304

### PR DESCRIPTION
**Pull request by jfz**
This PR fixes [issue #2796](https://github.com/apache/iceberg/issues/2796). It avoids invalid mark reset on retries by providing the input stream factory instead of original input stream to S3 API.

Root cause:

Currently, the underlying InputStream created from staging files is passed to S3 RequestBody and is reused by S3 client for every retry attempts; and S3 `mark` the underlying input stream with hard-coded of `read limit` of 128K at the beginning for once and `reset` it on every retry. In the case of a first attempt failed after reading more than 128K, the `mark` was invalidated because more than `read limit` amount of data has been read, and when then second attempt try to `reset` the invalidated mark, it breaks the mark/reset contract and throws exception with "Resetting to invalid mark". See java doc of `InputStream` for detail of mark/reset contract.

Steps to reproduce:

1. Get code of [s3mock](https://github.com/adobe/S3Mock) and run server with below change so that it fails the first request with 503 but succeeds on subsequent requests.

> FilestoreController.putObject:
> \+      LOG.info(" ######## " + count.incrementAndGet());
> \+      if(count.get() < 2) {
> \+        return ResponseEntity.status(503).build();
> \+      }
>        return ResponseEntity
>            .ok()
>            .eTag("\"" + s3Object.getEtag() + "\"")
> @@ -966,6 +970,7 @@ public class FileStoreController {
>            "Error persisting object.");
>      }
>    }
> \+  private static AtomicInteger count = new AtomicInteger(0);

2. Update unit test `TestS3OutputStream` as below and run `testInvalidMark`:

> \-  @ClassRule
> \-  public static final S3MockRule S3_MOCK_RULE = S3MockRule.builder().silent().build();
> \-
> \-  private final S3Client s3 = S3_MOCK_RULE.createS3ClientV2();
> \+//  @ClassRule
> \+//  public static final S3MockRule S3_MOCK_RULE = S3MockRule.builder().silent().build();
> \+//
> \+//  private final S3Client s3 = S3_MOCK_RULE.createS3ClientV2();
> \+  private final S3Client s3 = S3Client.builder()
> \+      .overrideConfiguration(ClientOverrideConfiguration.builder().retryPolicy(RetryPolicy.builder().numRetries(3).build()).build())
> \+      .region(Region.of("us\-east\-1"))
> \+      .credentialsProvider(
> \+              StaticCredentialsProvider.create(AwsBasicCredentials.create("foo", "bar")))
> \+      .endpointOverride(URI.create("http://localhost:" \+ 9090))
> \+      .httpClient(UrlConnectionHttpClient.builder().buildWithDefaults(AttributeMap.builder().build()))
> \+      .build();
> \+  @ Test()
> \+  public void testInvalidMark() {
> \+    byte[] data = randomData(129 * 1024);
> \+    writeAndVerify(s3mock, randomURI(), data, false);
> \+    ArgumentCaptor<PutObjectRequest> putObjectRequestArgumentCaptor =
> \+            ArgumentCaptor.forClass(PutObjectRequest.class);
> \+    verify(s3mock, times(1)).putObject(putObjectRequestArgumentCaptor.capture(),
> \+            (RequestBody) any());
> \+    checkPutObjectRequestContent(data, putObjectRequestArgumentCaptor);
> \+    checkTags(putObjectRequestArgumentCaptor);
> \+    reset(s3mock);
> \+  }

### CODE REVIEW

**Comment by jfz** [comment]
Another issue with `fromInputStream` is that it disables `close` of the original stream, and user is responsible for closing it outside of AWS SDK. In this particular case I think it's not too bad because SequenceInputStream closes the underlying input streams on iteration, only the BufferedInputStream is not closed but I think the buffer is GCed when this completes. cc @danielcweeks

**Comment by puneetzaroo** [comment]
Looks good

**Comment by stevenzwu** [comment]
@jfz can you refactor the manual testing in the description to proper unit tests?

**Comment by jfz** [comment]
> @jfz can you refactor the manual testing in the description to proper unit tests?

@stevenzwu Thanks for looking into this. The challenge to test retry is that it requires "S3Mock" code change and still tricky to get only a particular request to fail at first attempt and succeed later. The happen path (upload a s3 object with size over 128k) is well covered in existing tests though, the above example is actually simplified from an existing test.

**Review by rdblue — APPROVED** [review]

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 1]
```diff
@@ -389,7 +391,9 @@ private void completeUploads() {
       S3RequestUtil.configureEncryption(awsProperties, requestBuilder);
       S3RequestUtil.configurePermission(awsProperties, requestBuilder);
 
-      s3.putObject(requestBuilder.build(), RequestBody.fromInputStream(contentStream, contentLength));
+      s3.putObject(
+          requestBuilder.build(),
+          RequestBody.fromContentProvider(contentProvider, contentLength, Mimetype.MIMETYPE_OCTET_STREAM));
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:396]: Could this use `RequestBody.fromFile(f)` instead, like the `uploadParts` method does?

Comment (jfz) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:396]: The input stream here is a composition from multiple `stagingFiles`, is size of `stagingFiles` always 1 when it's not multipart? It's possible if that's the case, but I think the current way is a bit more elegant and future proof without that assumption.

Comment (rdblue) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:396]: Ah, you're right. We're concatenating them. Sounds good!

**Comment by rdblue** [comment]
Thanks, @jfz!

**Pull request merged**